turbopack/crates/turbopack-core/src/source_map/utils.rs

```diff
@@ -208,3 +267,128 @@ pub async fn fileify_source_map(
+                            "./current-dir-page.js",
+                            "../other%20route/page.js",
+                            // contains the file:// protocol/scheme
+                            url_root.join("absolute%20file%20url.js").unwrap().as_str(),
+                            // A server-relative path starting with `/`, potentially includes a
+                            // windows disk
+                            &format!("{}/server%20relative%20path.js", url_root.path()),
+                            // A scheme-relative path
+                            url_root
+                                .join("scheme%20relative%20path.js")
+                                .unwrap()
+                                .as_str()
+                                .strip_prefix("file:")
+                                .unwrap(),
+                            // non-file URLs are preserved
+                            "https://example.com/page%20path.js",
+                        ],
+                    )),
+                    // NOTE: the percent encoding here should NOT be decoded, as this is not part
+                    // of a `file://` URL
+                    &fs_root_path.join("app/source%20mapped/page.js").unwrap(),
+                )
+                .await?
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            )
+            .unwrap();
+
+            let prefix = format!("{SOURCE_URL_PROTOCOL}///[mock]");
+            assert_eq!(
+                resolved_source_map.sources,
+                Some(vec![
+                    Some(format!("{prefix}/app/source%20mapped/page.js")),
+                    Some(format!("{prefix}/app/source%20mapped/current-dir-page.js")),
+                    Some(format!("{prefix}/app/other route/page.js")),
+                    Some(format!("{prefix}/absolute file url.js")),
+                    Some(format!("{prefix}/server relative path.js")),
+                    Some(format!("{prefix}/scheme relative path.js")),
+                    Some("https://example.com/page%20path.js".to_owned()),
+                ])
+            );
+
+            // try with a `source_root`
+            let resolved_source_map: SourceMapJson = serde_json::from_str(
+                &resolve_source_map_sources(
+                    Some(&source_map_rope(
+                        // NOTE: these should get literally concated, a slash should NOT get added.
+                        Some("../source%20root%20"),
+                        ["page.js"],
+                    )),
+                    &fs_root_path.join("app/page.js").unwrap(),
+                )
+                .await?
+                .unwrap()
+                .to_str()
+                .unwrap(),
+            )
+            .unwrap();
+
+            assert_eq!(
+                resolved_source_map.sources,
+                Some(vec![Some(format!("{prefix}/source root page.js")),])
+            );
```

Comment:
https://tc39.es/ecma426/#sec-source-map-format: 
> The sources field is a list of original sources used by the [mappings field](https://tc39.es/ecma426/#json-mappings). Each entry is either a string that is a (potentially relative) [URL](https://tc39.es/ecma426/#external-whatwg-url) or null if the source name is not known.

`resolve_source_map_sources`  re-serializes that `SourceMapJson`, but these space would then have to be reencoded? Otherwise it's an url with an unencoded space in a path segment?